In [1]:
import graphlab as gl
from graphlab import SFrame
import pymysql
import numpy

Read in image files

In [2]:
image_sf = gl.image_analysis.load_images('./image_data')

This non-commercial license of GraphLab Create for academic use is assigned to workingjhy@gmail.com and will expire on April 29, 2018.


[INFO] graphlab.cython.cy_server: GraphLab Create v2.1 started. Logging: /tmp/graphlab_server_1494552822.log


Unsupported image format. Supported formats are JPG and PNG	 file: /Users/jihaoyu/Downloads/GraduateDesign/Code/image_data/.DS_Store

In [4]:
image_sf.column_names()

['path', 'image']

In [4]:
extrator = gl.feature_engineering.DeepFeatureExtractor(features='image', model = 'auto')

Downloading https://static.turi.com/products/graphlab-create/resources/models/python2.7/imagenet_model_iter45/objects.bin to /var/tmp/graphlab-HudsonJoe/2101/a9b19610-197b-4569-beb0-c865d2e7f197.bin

Failed to download https://static.turi.com/products/graphlab-create/resources/models/python2.7/imagenet_model_iter45/objects.bin: Timeout was reached

IOError: Unable to load model from https://static.turi.com/products/graphlab-create/resources/models/python2.7/imagenet_model_iter45: Cannot open https://static.turi.com/products/graphlab-create/resources/models/python2.7/imagenet_model_iter45/objects.bin for read. Fail to download from https://static.turi.com/products/graphlab-create/resources/models/python2.7/imagenet_model_iter45/objects.bin. Timeout was reached: unspecified iostream_category error: unspecified iostream_category error: unspecified iostream_category error: unspecified iostream_category error

In [7]:
extrator = extrator.fit(image_sf)

In [8]:
extr_model = extrator['model']

In [9]:
feature_sf = extrator.transform(image_sf)

Images being resized.

In [12]:
feature_sf[1]

{'deep_features.image': array('d', [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.10121744871139526, 0.0, 0.0, 0.7316427230834961, 0.0, 0.0, 0.9507333040237427, 0.0, 0.0, 1.0433980226516724, 0.0, 0.0, 0.0, 0.0, 0.5899171829223633, 0.0, 2.03592586517334, 0.9391565918922424, 0.0, 0.0, 0.0, 0.14870929718017578, 0.0, 0.0, 0.0, 0.0, 0.3000842332839966, 0.0, 0.0, 1.4000048637390137, 0.0, 0.0, 2.8663861751556396, 0.0, 1.941962718963623, 0.33416271209716797, 0.0, 0.0, 0.0, 1.294791340827942, 2.8249692916870117, 0.0, 0.0, 0.0, 0.0, 1.0534521341323853, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.2018623352050781, 1.8466746807098389, 0.0, 2.4671695232391357, 1.7704546451568604, 0.0, 0.0, 0.0, 0.0, 0.0, 1.4747507572174072, 0.0, 0.0, 0.24953830242156982, 0.0, 1.3037889003753662, 0.0, 0.0, 1.421032190322876, 0.0, 0.0, 0.0, 0.0, 0.24015462398529053, 0.0, 0.0, 2.3512954711914062, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0200424194335938, 2.1467225551605225, 0.0, 0.0, 1.968916654586792, 0.0, 0.0, 0.0, 0.0, 0.0, 3.439

In [14]:
features = list()
features.append('deep_features.image')

In [22]:
kmeans_model = gl.kmeans.create(feature_sf, num_clusters=15, features=features, max_iterations=10000)

Batch size is larger than the input dataset. Switching to an exact Kmeans method.

Choosing initial cluster centers with Kmeans++.

+---------------+-----------+

| Center number | Row index |

+---------------+-----------+

| 0             | 601       |

| 1             | 323       |

| 2             | 23        |

| 3             | 587       |

| 4             | 37        |

| 5             | 682       |

| 10            | 517       |

+---------------+-----------+

Starting kmeans model training.

Assigning points to initial cluster centers.

+-----------+-------------------------------+

| Iteration | Number of changed assignments |

+-----------+-------------------------------+

| 1         | 152                           |

| 2         | 69                            |

| 3         | 37                            |

| 4         | 19                            |

| 5         | 9                             |

| 6         | 6                             |

| 7         | 4                             |

| 8         | 2                             |

| 9         | 2                             |

| 10        | 2                             |

| 11        | 0                             |

+-----------+-------------------------------+

In [25]:
cluster_id = kmeans_model['cluster_id']
cluster_id

Columns:
	row_id	int
	cluster_id	int
	distance	float

Rows: 748

Data:
+--------+------------+---------------+
| row_id | cluster_id |    distance   |
+--------+------------+---------------+
|   0    |     2      | 51.9718475342 |
|   1    |     13     | 33.6384620667 |
|   2    |     4      | 35.2258911133 |
|   3    |     8      | 38.2093582153 |
|   4    |     8      | 36.5733146667 |
|   5    |     6      |  43.51304245  |
|   6    |     11     | 41.4294891357 |
|   7    |     14     | 39.9496192932 |
|   8    |     7      | 55.4590873718 |
|   9    |     8      | 51.5321426392 |
+--------+------------+---------------+
[748 rows x 3 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [21]:
print feature_sf[3]['path']
print feature_sf[5]['path']
print feature_sf[7]['path']

/Users/jihaoyu/Downloads/GraduateDesign/Code/image_data/10913a6539tb1xesx___0-item_pic.jpg
/Users/jihaoyu/Downloads/GraduateDesign/Code/image_data/0538e09cdctb2qmrx___2181865505.jpg
/Users/jihaoyu/Downloads/GraduateDesign/Code/image_data/0862e06ec8tb2c0ox___2181865505.jpg


In [29]:
feature_sf['cid_15'] = cluster_id['cluster_id']

In [30]:
feature_sf.column_names()

['path', 'image', 'deep_features.image', 'cid_15']

In [34]:
feature_sf[9]

{'cid_15': 8,
 'deep_features.image': array('d', [0.0, 0.0, 0.0, 0.0, 0.0, 1.1425590515136719, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 2.924454689025879, 0.7571301460266113, 0.0, 0.0, 0.0, 8.508834838867188, 0.0, 0.0, 0.0, 0.0, 0.0, 0.6272753477096558, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 2.53672456741333, 0.0, 0.4077697992324829, 0.5969405174255371, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.088393211364746, 0.0, 0.0, 0.39636754989624023, 1.4454739093780518, 0.0, 0.0, 0.0, 0.9843921065330505, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 5.560793876647949, 0.0, 5.179068565368652, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 3.3320250511169434, 0.0, 0.0, 0.0, 1.0090274810791016, 4.996869087219238, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.4191034436225891, 1.0460755825042725, 0.9390194416046143, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.15049755573272705, 0.0, 0.0, 0.0, 0.0, 0.0, 2.113708257675171, 0.4193938970565796, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.

In [33]:
feature_sf['img_path'] = feature_sf['path'].apply(lambda x: x.split('/')[-1])

In [35]:
feature_sf.column_names()

['path', 'image', 'deep_features.image', 'cid_15', 'img_path']

In [36]:
kmeans_less = gl.kmeans.create(feature_sf, num_clusters=5, features=features, max_iterations=10000)

Batch size is larger than the input dataset. Switching to an exact Kmeans method.

Choosing initial cluster centers with Kmeans++.

+---------------+-----------+

| Center number | Row index |

+---------------+-----------+

| 0             | 149       |

| 1             | 127       |

| 2             | 44        |

| 3             | 123       |

| 4             | 292       |

+---------------+-----------+

Starting kmeans model training.

Assigning points to initial cluster centers.

+-----------+-------------------------------+

| Iteration | Number of changed assignments |

+-----------+-------------------------------+

| 1         | 73                            |

| 2         | 34                            |

| 3         | 50                            |

| 4         | 58                            |

| 5         | 42                            |

| 6         | 27                            |

| 7         | 24                            |

| 8         | 18                            |

| 9         | 20                            |

| 10        | 39                            |

| 11        | 76                            |

| 12        | 26                            |

| 13        | 14                            |

| 14        | 10                            |

| 15        | 2                             |

| 16        | 1                             |

| 17        | 0                             |

+-----------+-------------------------------+

In [37]:
kmeans_less['cluster_id']

Columns:
	row_id	int
	cluster_id	int
	distance	float

Rows: 748

Data:
+--------+------------+---------------+
| row_id | cluster_id |    distance   |
+--------+------------+---------------+
|   0    |     2      |  51.229259491 |
|   1    |     4      | 40.1667671204 |
|   2    |     3      | 37.1395492554 |
|   3    |     0      | 48.5540847778 |
|   4    |     0      | 46.8035850525 |
|   5    |     4      | 46.1819572449 |
|   6    |     4      | 50.3599128723 |
|   7    |     4      | 42.6221237183 |
|   8    |     1      | 34.2808189392 |
|   9    |     0      | 60.9065208435 |
+--------+------------+---------------+
[748 rows x 3 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [44]:
feature_sf['cid_5'] = kmeans_less['cluster_id']['cluster_id']

In [40]:
kmeans_more = gl.kmeans.create(feature_sf, num_clusters=25, features=features, max_iterations=10000)

Batch size is larger than the input dataset. Switching to an exact Kmeans method.

Choosing initial cluster centers with Kmeans++.

+---------------+-----------+

| Center number | Row index |

+---------------+-----------+

| 0             | 151       |

| 1             | 162       |

| 2             | 633       |

| 3             | 729       |

| 4             | 378       |

| 5             | 723       |

| 10            | 32        |

+---------------+-----------+

Starting kmeans model training.

Assigning points to initial cluster centers.

+-----------+-------------------------------+

| Iteration | Number of changed assignments |

+-----------+-------------------------------+

| 1         | 124                           |

| 2         | 62                            |

| 3         | 38                            |

| 4         | 31                            |

| 5         | 16                            |

| 6         | 18                            |

| 7         | 14                            |

| 8         | 18                            |

| 9         | 9                             |

| 10        | 6                             |

| 11        | 2                             |

| 12        | 0                             |

+-----------+-------------------------------+

In [43]:
kmeans_more['cluster_id'].print_rows(num_rows = 100)

+--------+------------+---------------+
| row_id | cluster_id |    distance   |
+--------+------------+---------------+
|   0    |     11     | 48.2835083008 |
|   1    |     2      | 33.3306732178 |
|   2    |     0      | 39.0134963989 |
|   3    |     5      | 34.9269142151 |
|   4    |     22     | 34.2372589111 |
|   5    |     6      | 43.0008888245 |
|   6    |     12     |  41.553653717 |
|   7    |     7      |  39.60704422  |
|   8    |     20     | 44.3970985413 |
|   9    |     4      | 27.3413047791 |
|   10   |     11     |  48.309211731 |
|   11   |     1      | 28.6188716888 |
|   12   |     12     | 55.2747650146 |
|   13   |     0      | 18.4117927551 |
|   14   |     0      | 21.5263004303 |
|   15   |     8      | 48.7659187317 |
|   16   |     2      | 37.8827857971 |
|   17   |     21     | 43.7798194885 |
|   18   |     10     | 43.2991752625 |
|   19   |     14     | 45.5581474304 |
|   20   |     10     | 44.7593803406 |
|   21   |     15     |  37.501991272 |


In [45]:
feature_sf['cid_25'] = kmeans_more['cluster_id']['cluster_id']

In [46]:
feature_sf[0]

{'cid_15': 2,
 'cid_25': 11,
 'cid_5': 2,
 'deep_features.image': array('d', [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 4.0754923820495605, 0.1848122477531433, 0.0, 1.394418716430664, 0.0, 0.0, 0.0, 0.0, 6.6500935554504395, 0.0, 0.0, 0.0, 0.0, 0.0, 0.23811131715774536, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.6238279342651367, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.17537999153137207, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 3.1387293338775635, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.2435312271118164, 0.0, 0.0, 7.780476093292236, 0.9379912614822388, 0.0, 0.0, 0.0, 1.276609182357788, 0.0, 0.0, 5.059986591339111, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 4.60898494720459, 0.0, 0.0, 0.0, 0.0, 0.0, 3.9268481731414795, 0.0, 0.0, 0.0, 0.0, 0.0, 0.4439249038696289, 0.0, 0.0, 1.0342661142349243, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 7.985523223876953, 0.0, 2.927663564682007, 0.0, 0.0, 1.306870937347412, 0.0, 0.0, 0.0, 0.630825400352478, 2.92495584

In [47]:
config = {
    'host': 'localhost',
    'user': 'root',
    'password': '123456',
    'db': 'prod_schema',
    'charset': 'utf8mb4',
    'cursorclass': pymysql.cursors.DictCursor
}
def db_execute(config, sql):
    connection = pymysql.connect(**config)
    try:
        with connection.cursor() as cursor:
            cursor.execute(sql)
        connection.commit()
        return True
    finally:
        connection.close()

In [51]:
def update_cluster(config, row_data):
    img_path = row_data['img_path']
    cid_5 = row_data['cid_5']
    cid_15 = row_data['cid_15']
    cid_25 = row_data['cid_25']
    sql  = "update prod_schema.prod_img set cid_5 = %d, cid_15 = %d, cid_25 = %d \
where img_path = '%s'" % (cid_5, cid_15, cid_25, img_path)
    #print sql
    db_execute(config, sql)

In [50]:
update_cluster(config, feature_sf[0])

update prod_schema.prod_img set cid_5 = 2, cid_15 = 2, cid_25 = 11 where img_path = '099688ab56tb1ootx___0-item_pic.jpg'


In [52]:
for i in range(len(feature_sf)):
    row_data = feature_sf[i]
    update_cluster(config, row_data)

40 of 788 images are lost. Losing image data for 5.08%.
- Among 748 image data, **93** of them are assigned to the same cluster by the model of 5 clusters and the one of 15 clusters.
- Among 748 image data, **62** of them are assigned to the same cluster by the model of 15 clusters and the one of 25 clusters.
- None of these data belongs to a same cluster for all 3 model.